File: ports/go/src/gui/widgets/bcslayout.go
package widgets

import (
	"sync"
	core "github.com/robertpelloni/btk-go/ports/go/src/core/kernel"
	gui "github.com/robertpelloni/btk-go/ports/go/src/gui/kernel"
)

type BcsLayout struct {
	*core.BcsObject
	widgets []*gui.BcsWidget
	mu      sync.Mutex
}

func NewBcsLayout(parent *gui.BcsWidget) *BcsLayout {
	l := &BcsLayout{
		BcsObject: core.NewBcsObject(nil),
		widgets:   make([]*gui.BcsWidget, 0),
	}
	if parent != nil {
		l.SetParent(parent.BcsObject)
	}
	return l
}

func (l *BcsLayout) AddWidget(w *gui.BcsWidget) {
	l.mu.Lock()
	defer l.mu.Unlock()
	l.widgets = append(l.widgets, w)
}

func (l *BcsLayout) RemoveWidget(w *gui.BcsWidget) {
	l.mu.Lock()
	defer l.mu.Unlock()
	for i, widget := range l.widgets {
		if widget == w {
			l.widgets = append(l.widgets[:i], l.widgets[i+1:]...)
			break
		}
	}
}

func (l *BcsLayout) Widgets() []*gui.BcsWidget {
	l.mu.Lock()
	defer l.mu.Unlock()
	widgetsCopy := make([]*gui.BcsWidget, len(l.widgets))
	copy(widgetsCopy, l.widgets)
	return widgetsCopy
}

File: ports/java/src/gui/widgets/BcsLayout.java
package gui.widgets;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.locks.ReentrantLock;
import core.kernel.BcsObject;

public class BcsLayout extends BcsObject {
    private final List<BcsWidget> widgets;
    private final ReentrantLock lock = new ReentrantLock();

    public BcsLayout(BcsObject parent) {
        super(parent);
        this.widgets = new ArrayList<>();
    }

    public void addWidget(BcsWidget widget) {
        lock.lock();
        try {
            widgets.add(widget);
        } finally {
            lock.unlock();
        }
    }

    public void removeWidget(BcsWidget widget) {
        lock.lock();
        try {
            widgets.remove(widget);
        } finally {
            lock.unlock();
        }
    }

    public List<BcsWidget> getWidgets() {
        lock.lock();
        try {
            return new ArrayList<>(widgets);
        } finally {
            lock.unlock();
        }
    }
}
